### PR TITLE
Cleanup config parser

### PR DESCRIPTION
**Pull request by Damrod**
The config parser is a bit rough around the edges. This PR attempts to fix it.

### CODE REVIEW

File: srcs/app/config/analyser/Analyser.cpp [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
-	std::ostringstream str;
-	str << error << " in line " << line_;
-	lineerr_ = str.str();
+char	Analyser::SyntaxError::lineerror_[200] = {};
+
+Analyser::SyntaxError::SyntaxError(const std::string &error, size_t line) {
+	// if this didn't need to not throw, we would use std strings or string
+	// streams
+	snprintf(lineerror_, sizeof(lineerror_), "%s in line %zu",
+	error.c_str(), line);
```
Comment (gbudau) [inline srcs/app/config/analyser/Analyser.cpp:13]: Great idea to use snprintf. Thanks for the comments explaining why it is necessary.

File: config/default.conf [hunk 2]
```diff
@@ -54,7 +54,7 @@ server {
 
 	location	/cgi-bin/ {
 		client_max_body_size 100000000;
-		limit_except POST GET;
+		limit_except POST GET; #GET GET - duplicates work. Should they?
```
Comment (gbudau) [inline config/default.conf:57]: Duplicates work in nginx too. I think it is up to us to decide if we want to allow them or not. I don't have anything against them.

**Review by gbudau — APPROVED** [review]
Thank you! LGTM.

File: config/default.conf (hunk 2 above)
Comment (Damrod) [reply config/default.conf:57]: I've added some more unit tests

File: srcs/app/config/analyser/Analyser.cpp (hunk 1 above)
Comment (Damrod) [reply srcs/app/config/analyser/Analyser.cpp:13]: I made it by hand [here](https://github.com/42AGV/webserv/pull/118/commits/22d421eac53e3574bf419ebf9a7caebfb31d7a04), but then I told myself that it was ridiculous to do such kind of artisan micromanagement.

File: srcs/app/config/analyser/Lexer.cpp [hunk 3]
```diff
@@ -2,63 +2,72 @@
...
+		}
+		case '}' : {
+			ttype = Token::Type::T_SCOPE_CLOSE;
+			break;
+		}
+		default : {
+			throw Analyser::SyntaxError("Unexpected token", line_);
+		}
+	}
 	tmp[0] = type;
```
Comment (vdedios) [inline srcs/app/config/analyser/Lexer.cpp:41]: I've readen in many places that `switch` statements are evil but, as with everything, with moderation and common sense you can use all tools :smile: . I think it fits here :+1:

File: srcs/app/test/unit/config_analyser/ParserTest.cpp [hunk 4]
```diff
@@ -178,3 +174,91 @@ TEST_CASE("Testing the parser", "[parser]") {
...
+		std::ostringstream result;
+		std::vector<ServerConfig> cnf = tst.GetServersSettings();
+		result << cnf;
+	} catch(const std::exception &e) {
+		// std::cout << e.what();
+		REQUIRE(std::strcmp(e.what(),
+			"Keyword `listen' not allowed in context "
+			"`location' in line 42") == 0);
+	}
+}
```
Comment (vdedios) [inline srcs/app/test/unit/config_analyser/ParserTest.cpp:264]: Great you add some more tests. Thank you!

**Review by vdedios — APPROVED** [review]
Overall, good boy scouting. I think this PR improves readability! :rocket:

**Pull request merged**